mmdet/models/necks/bifpn.py

```diff
@@ -0,0 +1,326 @@
+            nodes = torch.stack(nodes, dim=-1)
+
+        nodes = torch.sum(nodes, dim=-1)
+
+        nodes = self.act_layer(nodes)
+
+        return self.fusion_convs(nodes)
+
+
+class BiFPNBlock(nn.Module):
+
+    def __init__(self, input_channels, num_backbone_features, num_outs,
+                 channels, weight_method, act_fn, separable_conv, epsilon,
+                 input_offsets, feature_reduction, reduction):
+        super().__init__()
+
+        weight_method_list = ["attn", "fast_attn", "sum"]
+        self.num_outs = num_outs
+
+        assert self.num_outs >= 3
+        assert weight_method in weight_method_list
+
+        self.nodes = nn.ModuleList()
+
+        for i in range(num_outs - 1):
+            self.nodes.append(BiFPNNode(input_channels[num_outs - i - 2],
+                                        channels,
+                                        weight_method,
+                                        act_fn,
+                                        separable_conv,
+                                        epsilon,
+                                        input_offsets[i],
+                                        reduction[i],
+                                        feature_reduction,
+                                        reduction))
+
+        for i in range(1, num_outs):
+            self.nodes.append(BiFPNNode(input_channels[i],
+                                        channels,
+                                        weight_method,
+                                        act_fn,
+                                        separable_conv,
+                                        epsilon,
+                                        input_offsets[num_outs + i - 2],
+                                        reduction[num_outs + i - 2],
+                                        feature_reduction,
+                                        reduction))
+
+    def forward(self, inputs):
+        output = list(inputs)
+
+        for node in self.nodes:
+            output.append(node(output))
+
+        return tuple(output[-self.num_outs::])
+
+
+@NECKS.register_module
+class BiFPN(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 input_indices,
+                 num_layers,
+                 num_outs,
+                 weight_method="fast_attn",
+                 act_fn="silu",
+                 separable_conv=True,
+                 epsilon=0.0001,
+                 base_reduction=8,
+                 reduction_ratio = 2.0):
+        super(BiFPN, self).__init__()
+        assert isinstance(in_channels, list)
+        self.num_backbone_features = len(in_channels)
+
+        assert self.num_backbone_features >= 2
+        assert num_outs - self.num_backbone_features >= 2
+
+        self.in_channels = in_channels
+        self.num_outs = num_outs
+
+        self.extra_convs = nn.ModuleList()
+
+        max_level = num_outs + self.num_backbone_features - 1
```

Comment:
Are you sure that it always True? What if num_outs=6, self.num_backbone_features  = 4

Reply:
Based on this [code](https://github.com/google/automl/blob/56815c9986ffd4b508fe1d68508e268d129715c1/efficientdet/keras/fpn_configs.py#L48). In this case, num_levels is num_outs (as they both represent the number of feature outputs) and self.num_backbone_features is min_level